# Implementação da Validação Cruzada na Base de Dados Breast Cancer
#### Usando validação cruzada para verificar a adaptação da rede neural artificial com os dados disponíveis.

### Importando as Bibliotecas Necessárias 

In [1]:
# importando as bibliotecas do python necessárias para construção e análise da rede neural

import pandas as pd
import numpy as np
import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import cross_val_score

Using TensorFlow backend.


### Separando os Atributos Previsores e Meta

In [2]:
# separando os atributos previsores e meta do dataframe
previsores = pd.read_csv('entradas_breast.csv', encoding = 'utf-8', sep = ',')
classe = pd.read_csv('saidas_breast.csv', encoding = 'utf-8', sep = ',')

### Configurando a Rede Neural

In [3]:
# criando a rede neural artificial
def neural_network_config():

  # criando o objeto 'classificador'
  classificador = Sequential()
  # adicionando a primeira rede oculta
  classificador.add(Dense(units = ((30 + 1) // 2 + 1), activation = 'relu', 
                          kernel_initializer = 'random_uniform', input_dim = 30))
  # adicionando uma segunda rede oculta
  classificador.add(Dense(units = ((30 + 1) // 2 + 1), activation = 'relu', 
                          kernel_initializer = 'random_uniform'))
  # adicionando uma rede de saída
  classificador.add(Dense(units = 1, activation = 'sigmoid'))
  # configurando a função de otimização no ajuste de pesos 'adam'
  otimizador = keras.optimizers.Adam(lr = 0.001, decay = 0.0001, clipvalue = 0.5)
  # ajustando mais parâmetros da rede neural artificial
  classificador.compile(optimizer = otimizador, loss = 'binary_crossentropy', metrics = ['binary_accuracy'])

  # função retorna o classificador configurado
  return classificador

In [4]:
# configurando os parâmetros finais de treinamento da rede neural
classificador = KerasClassifier(build_fn = neural_network_config, epochs = 100, batch_size = 10)

### Validação Cruzada

In [5]:
# adicionando a lista resultados os testes usando a validação cruzada
# será particionado a base de dados 10 vezes
resultados = cross_val_score(estimator = classificador, X = previsores, y = classe, cv = 10,
                             scoring = 'accuracy')

Epoch 1/100
512/512 [==============================] - 0s 767us/step - loss: 0.6243 - binary_accuracy: 0.6797
Epoch 2/100
512/512 [==============================] - 0s 112us/step - loss: 0.4505 - binary_accuracy: 0.7930
Epoch 3/100
512/512 [==============================] - 0s 97us/step - loss: 0.4800 - binary_accuracy: 0.8301
Epoch 4/100
512/512 [==============================] - 0s 101us/step - loss: 0.4508 - binary_accuracy: 0.8086
Epoch 5/100
512/512 [==============================] - 0s 98us/step - loss: 0.3777 - binary_accuracy: 0.8613
Epoch 6/100
512/512 [==============================] - 0s 105us/step - loss: 0.4357 - binary_accuracy: 0.8398
Epoch 7/100
512/512 [==============================] - 0s 97us/step - loss: 0.3755 - binary_accuracy: 0.8574
Epoch 8/100
512/512 [==============================] - 0s 100us/step - loss: 0.3435 - binary_accuracy: 0.8789
Epoch 9/100
512/512 [==============================] - 0s 93us/step - loss: 0.4192 - binary_accuracy: 0.8691
Epoch 10/100
5

In [6]:
# visualizando os resultados obtidos com a validação cruzada
print(resultados)

[0.8245614  0.89473684 0.92982456 0.78947368 0.89473684 0.89473684
 0.87719298 0.9122807  0.84210526 0.80357143]


Observe que para diferentes particionamentos da base de daods é obtido uma acurácia diferente. Desse modo, o valor médio desses valores tende a indicar uma acurácia mais real do que seria com mais e mais dados a serem usados para classificação.

In [7]:
# obtendo o resultado médio dos testes feitos usando validação cruzada
media = resultados.mean()
print(media)

0.8663220551378445


A acurácia precisa desse algoritmo é 86.63%. Esse valor é o mais real que podemos obter. Para analisar se os dados estão configurados na rede neural artificial de forma consistente, é necessário analisar o desvio padrão entre os dados.

In [8]:
# obtendo o desvio padrão dos resultados obtidos na validação cruzada
desvio = resultados.std()
print(desvio)

0.04567150227851201


Como o desvio entre os valores médios das bases de dados é baixo, descartamos a ocorrência de overfitting na base de dados.


### Alguma Dúvida? Entre em Contato Comigo:

- [Me envie um e-mail](mailto:alysson.barbosa@ee.ufcg.edu.br);